sdks/go/pkg/beam/core/runtime/harness/hooks.go

```diff
@@ -0,0 +1,112 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Hooks allow runners to tailor execution of the worker to allow for customization
+// of features used by the harness.
+//
+// Examples of customization:
+//
+// gRPC integration
+// session recording
+// profile recording
+//
+// Registration methods for hooks must be called prior to calling beam.Init()
+// Request methods for hooks must be called as part of building the pipeline
+// request for the runner's Execute method.
+
+package harness
+
+import (
+	"context"
+	"fmt"
+
+	fnpb "github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+)
+
+const hooksNamespaceKey = "__hooks:"
+
+// HookKey returns a hook registration key for the specified input.
+func HookKey(in string) string {
```

Comment:
I don't think we need to expose this key. Instead, we could expose EnableHook(string) / DisableHook(string) that set the appropriate global option.

Reply:
That's cleaner. Implementing that.

Reply:
Actually, I had considered that. In my notes, I went with this approach because if a hook wanted to pass additional data via options, having a "subtree" in the options structure allowed a clean way to do that. If we did the EnableHook() approach, we wouldn't be solving that need for the user. Since it's somewhat tightly coupled, I went this direction.